File: GhostChat-Android/app/src/main/java/com/ghost/chat/core/security/BiometricAuthService.kt
package com.ghost.chat.core.security

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.ghost.chat.core.crypto.IdentityKeyService
import com.ghost.chat.core.storage.DatabaseService
import java.io.File
import java.security.KeyStore
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec

/// Biometric + PIN authentication service
/// PIN code is stored as PBKDF2-HMAC-SHA256 hash with random salt in KeystoreService
/// Brute force protection: lockout after 5 attempts, wipe after 10
/// Biometric auth is only available when a PIN code is already set
object BiometricAuthService {

    private const val BIOMETRIC_KEY_ALIAS = "ghost_biometric_key"
    private const val KEY_BIOMETRIC_ENABLED = "biometric_enabled"
    private const val KEY_PIN_HASH = "pin_code_hash"
    private const val KEY_PIN_LENGTH = "pin_code_length"
    private const val KEY_AUTOLOCK_SECONDS = "autolock_seconds"
    private const val KEY_BACKGROUND_TS = "background_timestamp"

    private lateinit var appContext: android.content.Context

    var isUnlocked: Boolean by mutableStateOf(true)
        private set

    // Cold start: determine initial lock state
    // Must be called after KeystoreService.init()
    fun initLockState(context: android.content.Context) {
        appContext = context.applicationContext
        val pinSet = KeystoreService.loadString(KEY_PIN_HASH) != null
        val bioEnabled = KeystoreService.loadBool(KEY_BIOMETRIC_ENABLED, false)
        if (pinSet || bioEnabled) {
            val autoLock = KeystoreService.loadInt(KEY_AUTOLOCK_SECONDS, 0)
            if (autoLock > 0) {
                // Timer-based: check if background timestamp is within the window
                val tsStr = KeystoreService.loadString(KEY_BACKGROUND_TS)
                val ts = tsStr?.toLongOrNull()
                if (ts != null) {
                    val elapsed = (System.currentTimeMillis() / 1000) - ts
                    isUnlocked = elapsed < autoLock
                } else {
                    isUnlocked = false
                }
                KeystoreService.delete(KEY_BACKGROUND_TS)
            } else {
                // Instant lock — always lock on cold start
                isUnlocked = false
            }
        }
        // else: no PIN and no biometric → isUnlocked stays true
    }

    var isEnabled: Boolean
        get() = KeystoreService.loadBool(KEY_BIOMETRIC_ENABLED, false)
        set(value) = KeystoreService.saveBool(value, KEY_BIOMETRIC_ENABLED)

    /** Whether a PIN code has been set */
    val isPinSet: Boolean
        get() = KeystoreService.loadString(KEY_PIN_HASH) != null

    /** PIN length: 4 or 6 digits */
    var pinLength: Int
        get() = KeystoreService.loadInt(KEY_PIN_LENGTH, 4)
        set(value) = KeystoreService.saveInt(value, KEY_PIN_LENGTH)

    /** Auto-lock timer in seconds (0 = instant) */
    var autoLockSeconds: Int
        get() = KeystoreService.loadInt(KEY_AUTOLOCK_SECONDS, 0)
        set(value) = KeystoreService.saveInt(value, KEY_AUTOLOCK_SECONDS)

    /** PBKDF2-HMAC-SHA256 key derivation (600k iterations, OWASP recommendation) */
    private fun pbkdf2(pin: String, salt: ByteArray, iterations: Int = 600_000): ByteArray {
        val spec = PBEKeySpec(pin.toCharArray(), salt, iterations, 256)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded
    }

    /** Set a new PIN code (stores PBKDF2 hash with random salt) */
    fun setPin(pin: String) {
        val salt = ByteArray(16).also { SecureRandom().nextBytes(it) }
        val derived = pbkdf2(pin, salt)
        val stored = salt + derived  // 16 bytes salt + 32 bytes key
        KeystoreService.saveString(
            stored.joinToString("") { "%02x".format(it) },
            KEY_PIN_HASH
        )
        resetPinAttempts()
    }

    /** Verify PIN code against stored PBKDF2 hash. Sets isUnlocked=true on match. */
    fun verifyPin(pin: String): Boolean {
        // Brute force lockout check
        if (isPinLocked()) return false

        val storedHex = KeystoreService.loadString(KEY_PIN_HASH) ?: return false
        val stored = storedHex.chunked(2).map { it.toInt(16).toByte() }.toByteArray()

        // Support both old SHA-256 (32 bytes) and new PBKDF2 (48 bytes = 16 salt + 32 hash)
        val matches = if (stored.size == 48) {
            // New format: PBKDF2 with salt
            val salt = stored.copyOfRange(0, 16)
            val storedHash = stored.copyOfRange(16, 48)
            val derived = pbkdf2(pin, salt)

            // Constant-time comparison
            var diff = 0
            for (i in derived.indices) {
                diff = diff or (derived[i].toInt() xor storedHash[i].toInt())
            }
            diff == 0
        } else {
            // Legacy SHA-256 format — migrate on successful verify
            val digest = java.security.MessageDigest.getInstance("SHA-256")
            val inputHash = digest.digest(pin.toByteArray(Charsets.UTF_8))
                .joinToString("") { "%02x".format(it) }
            val legacyMatch = storedHex == inputHash
            if (legacyMatch) {
                // Migrate to PBKDF2
                setPin(pin)
            }
            legacyMatch
        }

        if (matches) {
            resetPinAttempts()
            isUnlocked = true
        } else {
            onPinFailed()
        }
        return matches
    }

    /** Remove PIN code and disable biometric auth */
    fun removePin() {
        KeystoreService.delete(KEY_PIN_HASH)
        resetPinAttempts()
        isEnabled = false
        isUnlocked = true
    }

    // MARK: - Brute Force Protection

    private fun isPinLocked(): Boolean {
        val lockedUntil = KeystoreService.loadLong("pin_locked_until", 0)
        if (lockedUntil > System.currentTimeMillis()) return true
        return false
    }

    private fun onPinFailed() {
        val attempts = KeystoreService.loadInt("pin_failed_attempts", 0) + 1
        KeystoreService.saveInt(attempts, "pin_failed_attempts")

        when {
            attempts >= 10 -> {
                // Panic wipe — уничтожение всех данных после 10 неудачных попыток
                // 1. Destroy and close the SQLCipher database
                try {
                    val dbService = DatabaseService.getInstance(appContext)
                    dbService.destroyAll()
                    dbService.close()
                } catch (_: Exception) {}
                // 2. Delete the DB file from disk
                try {
                    appContext.getDatabasePath("ghost_chat.db")?.let { dbFile ->
                        if (dbFile.exists()) dbFile.delete()
                    }
                } catch (_: Exception) {}
                // 3. Destroy identity keys
                IdentityKeyService.destroy()
                // 4. Clear all Keystore data (PIN hash, settings, DB key, etc.)
                KeystoreService.clear()
                // 5. Reset lock state — app is wiped, nothing to protect
                isUnlocked = true
            }
            attempts >= 8 -> {
                // 5 минут блокировки
                KeystoreService.saveLong(
                    System.currentTimeMillis() + 5 * 60 * 1000,
                    "pin_locked_until"
                )
            }
            attempts >= 5 -> {
                // 30 секунд блокировки
                KeystoreService.saveLong(
                    System.currentTimeMillis() + 30 * 1000,
                    "pin_locked_until"
                )
            }
        }
    }

    private fun resetPinAttempts() {
        KeystoreService.saveInt(0, "pin_failed_attempts")
        KeystoreService.saveLong(0, "pin_locked_until")
    }

    // MARK: - Biometric-bound Keystore Key (hardware-enforced)

    /** Get or create AES key bound to biometric authentication in Android Keystore.
     *  setUserAuthenticationRequired(true) ensures the key can only be used after
     *  successful biometric auth — hardware-enforced, cannot be bypassed on rooted devices. */
    private fun getBiometricKey(): SecretKey {
        val keyStore = KeyStore.getInstance("AndroidKeyStore")
        keyStore.load(null)

        // Return existing key if available
        keyStore.getKey(BIOMETRIC_KEY_ALIAS, null)?.let { return it as SecretKey }

        // Create new biometric-bound key
        val keyGen = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
        keyGen.init(
            KeyGenParameterSpec.Builder(
                BIOMETRIC_KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setUserAuthenticationRequired(true)
                .setInvalidatedByBiometricEnrollment(true)
                .build()
        )
        return keyGen.generateKey()
    }

    /** Create CryptoObject with biometric-bound cipher.
     *  Returns null if key was invalidated (e.g. new biometric enrolled). */
    private fun createCryptoObject(): BiometricPrompt.CryptoObject? {
        return try {
            val key = getBiometricKey()
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, key)
            BiometricPrompt.CryptoObject(cipher)
        } catch (e: Exception) {
            // Key invalidated (new biometric enrolled) or other Keystore error —
            // delete the old key so a fresh one is created next time
            try {
                val keyStore = KeyStore.getInstance("AndroidKeyStore")
                keyStore.load(null)
                keyStore.deleteEntry(BIOMETRIC_KEY_ALIAS)
            } catch (_: Exception) { }
            null
        }
    }

    /** Check if BIOMETRIC_STRONG is available (fingerprint/face with crypto support) */
    private fun canUseBiometricStrong(context: Context): Boolean {
        val biometricManager = BiometricManager.from(context)
        return biometricManager.canAuthenticate(
            BiometricManager.Authenticators.BIOMETRIC_STRONG
        ) == BiometricManager.BIOMETRIC_SUCCESS
    }

    /** Check if biometric auth is available on device */
    fun isAvailable(context: Context): Boolean {
        val biometricManager = BiometricManager.from(context)
        return biometricManager.canAuthenticate(
            BiometricManager.Authenticators.BIOMETRIC_STRONG or
                    BiometricManager.Authenticators.DEVICE_CREDENTIAL
        ) == BiometricManager.BIOMETRIC_SUCCESS
    }

    /** Authenticate with biometrics (hardware-enforced CryptoObject) or DEVICE_CREDENTIAL fallback */
    fun authenticate(
        activity: FragmentActivity,
        title: String,
        subtitle: String,
        onSuccess: () -> Unit,
        onError: (String) -> Unit
    ) {
        val executor = ContextCompat.getMainExecutor(activity)

        val callback = object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                // result.cryptoObject is non-null for BIOMETRIC_STRONG path —
                // proves the hardware TEE actually authenticated the user
                isUnlocked = true
                onSuccess()
            }

            override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                if (errorCode != BiometricPrompt.ERROR_USER_CANCELED &&
                    errorCode != BiometricPrompt.ERROR_NEGATIVE_BUTTON
                ) {
                    onError(errString.toString())
                }
            }

            override fun onAuthenticationFailed() {
                // Not an error — user can retry
            }
        }

        val biometricPrompt = BiometricPrompt(activity, executor, callback)

        // Try BIOMETRIC_STRONG with CryptoObject (hardware-enforced, cannot bypass on rooted devices)
        if (canUseBiometricStrong(activity)) {
            val cryptoObject = createCryptoObject()
            if (cryptoObject != null) {
                val promptInfo = BiometricPrompt.PromptInfo.Builder()
                    .setTitle(title)
                    .setSubtitle(subtitle)
                    .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
                    .setNegativeButtonText("PIN")
                    .build()

                biometricPrompt.authenticate(promptInfo, cryptoObject)
                return
            }
        }

        // Fallback: DEVICE_CREDENTIAL (PIN/pattern/password — no CryptoObject allowed)
        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .setSubtitle(subtitle)
            .setAllowedAuthenticators(
                BiometricManager.Authenticators.BIOMETRIC_STRONG or
                        BiometricManager.Authenticators.DEVICE_CREDENTIAL
            )
            .build()

        biometricPrompt.authenticate(promptInfo)
    }

    /** Record background timestamp (for auto-lock timer) */
    fun didEnterBackground() {
        if (!isPinSet && !isEnabled) return

        if (autoLockSeconds == 0) {
            // Instant lock
            isUnlocked = false
        } else {
            // Record timestamp for delayed lock
            KeystoreService.saveString(
                "${System.currentTimeMillis() / 1000}",
                KEY_BACKGROUND_TS
            )
        }
    }

    /** Check auto-lock timer on foreground return */
    fun didEnterForeground() {
        if (!isPinSet && !isEnabled) return
        if (autoLockSeconds == 0) return // instant lock already handled

        val tsStr = KeystoreService.loadString(KEY_BACKGROUND_TS) ?: return
        val ts = tsStr.toLongOrNull() ?: return
        val elapsed = (System.currentTimeMillis() / 1000) - ts
        if (elapsed >= autoLockSeconds) {
            isUnlocked = false
        }
        KeystoreService.delete(KEY_BACKGROUND_TS)
    }

    /** Lock immediately */
    fun lock() {
        if (isPinSet || isEnabled) {
            isUnlocked = false
        }
    }

    /** Toggle biometric auth (requires PIN to be set first) */
    fun toggle(
        activity: FragmentActivity,
        title: String,
        subtitle: String,
        onResult: (Boolean) -> Unit
    ) {
        if (!isPinSet) {
            // Cannot enable biometric without PIN
            onResult(false)
            return
        }

        if (isEnabled) {
            isEnabled = false
            onResult(false)
        } else {
            authenticate(activity, title, subtitle,
                onSuccess = {
                    isEnabled = true
                    onResult(true)
                },
                onError = { onResult(false) }
            )
        }
    }
}

File: GhostChat-iOS/GhostChat/Core/Security/BiometricAuthService.swift
import Foundation
import LocalAuthentication
import CryptoKit
import CommonCrypto

/// Face ID / Touch ID + PIN code authentication service
/// Settings stored in Keychain (not UserDefaults) for security
@MainActor
final class BiometricAuthService: ObservableObject {

    // MARK: - Published State

    @Published var isUnlocked = false
    @Published private(set) var isEnabled: Bool
    @Published private(set) var isPinSet: Bool
    @Published var pinLength: Int {
        didSet {
            if let data = "\(pinLength)".data(using: .utf8) {
                KeychainService.save(data, forKey: Self.pinLengthKey)
            }
        }
    }
    @Published var autoLockSeconds: Int {
        didSet {
            if let data = "\(autoLockSeconds)".data(using: .utf8) {
                KeychainService.save(data, forKey: Self.autoLockKey)
            }
        }
    }

    // MARK: - Constants

    private static let keychainKey = "biometric_enabled"
    private static let pinHashKey = "app_pin_hash"
    private static let pinLengthKey = "app_pin_length"
    private static let autoLockKey = "app_autolock_seconds"
    private static let backgroundTimestampKey = "app_background_ts"
    private static let failedAttemptsKey = "pin_failed_attempts"
    private static let lockedUntilKey = "pin_locked_until"

    // MARK: - Brute Force Protection

    private var failedPinAttempts: Int {
        get {
            guard let data = KeychainService.load(forKey: Self.failedAttemptsKey),
                  let str = String(data: data, encoding: .utf8),
                  let count = Int(str) else { return 0 }
            return count
        }
        set {
            if let data = "\(newValue)".data(using: .utf8) {
                KeychainService.save(data, forKey: Self.failedAttemptsKey)
            }
        }
    }

    private var pinLockedUntil: Date? {
        get {
            guard let data = KeychainService.load(forKey: Self.lockedUntilKey),
                  let str = String(data: data, encoding: .utf8),
                  let ts = Double(str) else { return nil }
            return Date(timeIntervalSince1970: ts)
        }
        set {
            if let date = newValue {
                let ts = "\(date.timeIntervalSince1970)"
                if let data = ts.data(using: .utf8) {
                    KeychainService.save(data, forKey: Self.lockedUntilKey)
                }
            } else {
                KeychainService.delete(forKey: Self.lockedUntilKey)
            }
        }
    }

    // MARK: - Init

    init() {
        // Load biometric setting from Keychain
        if let data = KeychainService.load(forKey: Self.keychainKey),
           let value = String(data: data, encoding: .utf8) {
            self.isEnabled = value == "1"
        } else {
            self.isEnabled = false
        }

        // Check if PIN is set
        self.isPinSet = KeychainService.load(forKey: Self.pinHashKey) != nil

        // Load PIN length (default 4)
        if let data = KeychainService.load(forKey: Self.pinLengthKey),
           let str = String(data: data, encoding: .utf8),
           let len = Int(str) {
            self.pinLength = len
        } else {
            self.pinLength = 4
        }

        // Load auto-lock timer (default 0 = instant)
        if let data = KeychainService.load(forKey: Self.autoLockKey),
           let str = String(data: data, encoding: .utf8),
           let secs = Int(str) {
            self.autoLockSeconds = secs
        } else {
            self.autoLockSeconds = 0
        }

        // Determine initial lock state
        if !isPinSet && !isEnabled {
            // No security configured — start unlocked
            isUnlocked = true
        } else if autoLockSeconds > 0 {
            // Timer-based lock: check if background timestamp is within the window
            // This enables "stay unlocked on cold start if timer hasn't expired"
            if let data = KeychainService.load(forKey: Self.backgroundTimestampKey),
               let str = String(data: data, encoding: .utf8),
               let ts = Int(str) {
                let elapsed = Int(Date().timeIntervalSince1970) - ts
                if elapsed < autoLockSeconds {
                    isUnlocked = true
                }
                // else: timer expired → isUnlocked stays false
            }
            // Clean up timestamp
            KeychainService.delete(forKey: Self.backgroundTimestampKey)
        }
        // else: instant lock (autoLockSeconds == 0) → isUnlocked stays false
    }

    // MARK: - Biometric Type Detection

    enum BiometricType {
        case faceID
        case touchID
        case none
    }

    static var biometricType: BiometricType {
        let context = LAContext()
        var error: NSError?
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
            return .none
        }

        switch context.biometryType {
        case .faceID:
            return .faceID
        case .touchID:
            return .touchID
        case .opticID:
            return .faceID // Vision Pro — treat as Face ID
        case .none:
            return .none
        @unknown default:
            return .none
        }
    }

    static var isAvailable: Bool {
        biometricType != .none
    }

    var biometricName: String {
        switch Self.biometricType {
        case .faceID:
            return String(localized: "biometric.faceid")
        case .touchID:
            return String(localized: "biometric.touchid")
        case .none:
            return ""
        }
    }

    // MARK: - PIN Management

    /// Set or change PIN code (stored as PBKDF2-HMAC-SHA256 hash + salt in Keychain)
    func setPin(_ pin: String) {
        let salt = Data((0..<16).map { _ in UInt8.random(in: 0...255) })
        guard let derivedKey = pbkdf2(pin: pin, salt: salt) else { return }
        let stored = salt + derivedKey  // 16 bytes salt + 32 bytes derived key
        KeychainService.save(stored, forKey: Self.pinHashKey)
        isPinSet = true
        // Reset brute force counters on PIN change
        failedPinAttempts = 0
        pinLockedUntil = nil
    }

    /// PBKDF2-HMAC-SHA256 key derivation (600k rounds)
    private func pbkdf2(pin: String, salt: Data, rounds: Int = 600_000) -> Data? {
        let pinData = Data(pin.utf8)
        var derivedKey = Data(count: 32)
        let result = derivedKey.withUnsafeMutableBytes { derivedKeyBytes in
            salt.withUnsafeBytes { saltBytes in
                pinData.withUnsafeBytes { pinBytes in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        pinBytes.baseAddress?.assumingMemoryBound(to: Int8.self),
                        pinData.count,
                        saltBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                        UInt32(rounds),
                        derivedKeyBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        32
                    )
                }
            }
        }
        return result == kCCSuccess ? derivedKey : nil
    }

    /// Verify entered PIN against stored PBKDF2 hash with brute force protection
    func verifyPin(_ pin: String) -> Bool {
        // Check brute force lockout
        if let lockedUntil = pinLockedUntil, Date() < lockedUntil {
            return false
        }

        guard let stored = KeychainService.load(forKey: Self.pinHashKey) else { return false }

        // Support both old (32-byte SHA256) and new (48-byte PBKDF2) format
        let match: Bool
        if stored.count == 48 {
            // New format: 16 bytes salt + 32 bytes PBKDF2 hash
            let salt = Data(stored.prefix(16))
            let storedHash = Data(stored.suffix(32))
            guard let derivedKey = pbkdf2(pin: pin, salt: salt) else { return false }
            // Constant-time comparison
            var result: UInt8 = 0
            for (a, b) in zip(derivedKey, storedHash) { result |= a ^ b }
            match = result == 0
        } else {
            // Legacy format: 32-byte SHA256 — migrate on success
            let hash = Data(SHA256.hash(data: Data(pin.utf8)))
            match = hash == stored
            if match {
                // Auto-migrate to PBKDF2 format
                setPin(pin)
            }
        }

        if match {
            isUnlocked = true
            failedPinAttempts = 0
            pinLockedUntil = nil
        } else {
            let attempts = failedPinAttempts + 1
            failedPinAttempts = attempts
            if attempts >= 10 {
                // PANIC: 10 failed attempts — wipe all data
                panicWipe()
            } else if attempts >= 8 {
                // Lock for 5 minutes
                pinLockedUntil = Date().addingTimeInterval(300)
            } else if attempts >= 5 {
                // Lock for 30 seconds
                pinLockedUntil = Date().addingTimeInterval(30)
            }
        }
        return match
    }

    /// Remaining lockout seconds (for UI display)
    var pinLockoutRemaining: TimeInterval {
        guard let lockedUntil = pinLockedUntil else { return 0 }
        return max(0, lockedUntil.timeIntervalSinceNow)
    }

    /// Emergency wipe — 10 failed PIN attempts
    private func panicWipe() {
        // Destroy database + encryption key + identity key
        DatabaseService.destroy()
        // Clear all Keychain data
        KeychainService.delete(forKey: Self.pinHashKey)
        KeychainService.delete(forKey: Self.keychainKey)
        KeychainService.delete(forKey: Self.pinLengthKey)
        KeychainService.delete(forKey: Self.autoLockKey)
        KeychainService.delete(forKey: Self.failedAttemptsKey)
        KeychainService.delete(forKey: Self.lockedUntilKey)
        KeychainService.delete(forKey: Self.backgroundTimestampKey)
        // Reset state
        isPinSet = false
        isEnabled = false
        isUnlocked = true
        failedPinAttempts = 0
        pinLockedUntil = nil
    }

    /// Remove PIN code and disable biometric (biometric requires PIN)
    func removePin() {
        KeychainService.delete(forKey: Self.pinHashKey)
        isPinSet = false
        // Biometric requires PIN — disable it too
        setEnabled(false)
        isUnlocked = true
    }

    // MARK: - Authentication

    /// Authenticate with biometrics (or device passcode as fallback)
    func authenticate() async -> Bool {
        let context = LAContext()
        context.localizedFallbackTitle = "" // Hide "Enter Password" to force biometric
        context.localizedCancelTitle = nil

        var error: NSError?

        // Try biometrics first
        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
            // Biometrics unavailable — try device passcode as fallback
            return await authenticateWithPasscode()
        }

        do {
            let success = try await context.evaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                localizedReason: String(localized: "biometric.reason")
            )
            if success {
                isUnlocked = true
            }
            return success
        } catch {
            // Biometric failed (e.g. face not recognized) — try passcode
            if let laError = error as? LAError,
               laError.code == .biometryLockout || laError.code == .biometryNotAvailable {
                return await authenticateWithPasscode()
            }
            return false
        }
    }

    /// Fallback: device passcode authentication
    private func authenticateWithPasscode() async -> Bool {
        let context = LAContext()

        var error: NSError?
        guard context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error) else {
            return false
        }

        do {
            let success = try await context.evaluatePolicy(
                .deviceOwnerAuthentication,
                localizedReason: String(localized: "biometric.reason")
            )
            if success {
                isUnlocked = true
            }
            return success
        } catch {
            return false
        }
    }

    // MARK: - Toggle

    /// Enable or disable biometric lock (requires PIN to be set first)
    func setEnabled(_ enabled: Bool) {
        // Biometric requires PIN to be set first
        if enabled && !isPinSet { return }

        isEnabled = enabled
        let value = enabled ? "1" : "0"
        if let data = value.data(using: .utf8) {
            KeychainService.save(data, forKey: Self.keychainKey)
        }

        // If disabling and no PIN, unlock immediately
        if !enabled && !isPinSet {
            isUnlocked = true
        }
    }

    /// Save background timestamp (for auto-lock timer).
    /// Called on .inactive — harmless if .background also fires (it will overwrite).
    func saveBackgroundTimestamp() {
        guard isPinSet || isEnabled else { return }
        guard autoLockSeconds > 0 else { return }

        let ts = "\(Int(Date().timeIntervalSince1970))"
        if let data = ts.data(using: .utf8) {
            KeychainService.save(data, forKey: Self.backgroundTimestampKey)
        }
    }

    /// Record background timestamp (for auto-lock timer)
    func didEnterBackground() {
        guard isPinSet || isEnabled else { return }

        if autoLockSeconds == 0 {
            // Instant lock
            isUnlocked = false
        } else {
            // Record timestamp for delayed lock
            saveBackgroundTimestamp()
        }
    }

    /// Check auto-lock timer on foreground return
    func didEnterForeground() {
        guard isPinSet || isEnabled else { return }
        guard autoLockSeconds > 0 else { return } // instant lock already handled

        if let data = KeychainService.load(forKey: Self.backgroundTimestampKey),
           let str = String(data: data, encoding: .utf8),
           let ts = Int(str) {
            let elapsed = Int(Date().timeIntervalSince1970) - ts
            if elapsed >= autoLockSeconds {
                isUnlocked = false
            }
        }
        // Clean up timestamp
        KeychainService.delete(forKey: Self.backgroundTimestampKey)
    }

    /// Lock the app immediately
    func lock() {
        guard isPinSet || isEnabled else { return }
        isUnlocked = false
    }

    /// Re-read Keychain state (used after first-launch cleanup)
    func refreshState() {
        if let data = KeychainService.load(forKey: Self.keychainKey),
           let value = String(data: data, encoding: .utf8) {
            isEnabled = value == "1"
        } else {
            isEnabled = false
        }
        isPinSet = KeychainService.load(forKey: Self.pinHashKey) != nil

        if let data = KeychainService.load(forKey: Self.pinLengthKey),
           let str = String(data: data, encoding: .utf8),
           let len = Int(str) {
            pinLength = len
        } else {
            pinLength = 4
        }

        if let data = KeychainService.load(forKey: Self.autoLockKey),
           let str = String(data: data, encoding: .utf8),
           let secs = Int(str) {
            autoLockSeconds = secs
        } else {
            autoLockSeconds = 0
        }

        if !isPinSet && !isEnabled {
            isUnlocked = true
        }
    }
}
